lib/Interpreter/ValuePrinter.cpp

```diff
@@ -151,52 +146,59 @@ bool canParseTypeName(cling::Interpreter& Interp, llvm::StringRef typenam) {
 }
 #endif
 
+static std::string printDeclType(const clang::QualType& QT,
+                                 const clang::NamedDecl* D) {
+  if (!QT.hasQualifiers())
+    return D->getQualifiedNameAsString();
+  return QT.getQualifiers().getAsString() + " " + D->getQualifiedNameAsString();
+}
+
 static std::string printQualType(clang::ASTContext& Ctx, clang::QualType QT) {
   using namespace clang;
-  std::ostringstream strm;
-  QualType QTNonRef = QT.getNonReferenceType();
-  std::string ValueTyStr;
-  if (const TypedefType *TDTy = dyn_cast<TypedefType>(QTNonRef))
-    ValueTyStr = TDTy->getDecl()->getQualifiedNameAsString();
-  else if (const TagType *TTy = dyn_cast<TagType>(QTNonRef))
-    ValueTyStr = TTy->getDecl()->getQualifiedNameAsString();
+  const QualType QTNonRef = QT.getNonReferenceType();
+  const QualType QTCanon = QTNonRef.getCanonicalType();
 
-  if (ValueTyStr.empty())
-    ValueTyStr = cling::utils::TypeName::GetFullyQualifiedName(QTNonRef, Ctx);
-  else if (QTNonRef.hasQualifiers())
-    ValueTyStr = QTNonRef.getQualifiers().getAsString() + " " + ValueTyStr;
+  std::string ValueTyStr("(");
+  if ((isa<TypedefType>(QTNonRef) || QTCanon->isBuiltinType())
+      && !QTNonRef->isFunctionPointerType() && !QTNonRef->isMemberPointerType())
+    ValueTyStr += QTCanon.getAsString(Ctx.getPrintingPolicy());
+  else if (const TagType *TTy = dyn_cast<TagType>(QTNonRef))
+    ValueTyStr += printDeclType(QTNonRef, TTy->getDecl());
+  else if (const RecordType *TRy = dyn_cast<RecordType>(QTNonRef))
+    ValueTyStr += printDeclType(QTNonRef, TRy->getDecl());
+  else
+    ValueTyStr += cling::utils::TypeName::GetFullyQualifiedName(QTNonRef, Ctx);
 
-  strm << "(";
-  strm << ValueTyStr;
   if (QT->isReferenceType())
-    strm << " &";
-  strm << ")";
-  return strm.str();
+    ValueTyStr += " &";
+
+  return ValueTyStr + ")";
 }
 
 } // anonymous namespace
 
 template<typename T>
 static std::string executePrintValue(const Value &V, const T &val) {
-  // don't use std::stringstream, since it doesn't prepend '0x'
-  // in front of hexadecimal values when streaming pointer values
-  std::string strval;
-  llvm::raw_string_ostream printValueSS(strval);
-  printValueSS << "cling::printValue(";
-  printValueSS << getTypeString(V);
-  printValueSS << (const void *) &val;
-  printValueSS << ");";
-
   Interpreter *Interp = V.getInterpreter();
   Value printValueV;
 
   {
+    // don't use std::stringstream, since it doesn't prepend '0x'
+    // in front of hexadecimal values when streaming pointer values
+
+    llvm::SmallString<512> Buf;
+    llvm::raw_svector_ostream Strm(Buf);
+    Strm << "cling::printValue(";
+    Strm << getTypeString(V);
+    Strm << (const void*) &val;
+    Strm << ");";
```

Comment:
It could drop the line count by 1, but as there is already a local stream would be less efficient.